coral-hive/src/main/java/com/linkedin/coral/transformers/CoralRelToSqlNodeConverter.java

```diff
@@ -345,4 +348,43 @@ private SqlNode generateRightChildForSqlJoinWithLateralViews(BiRel e, Result rig
 
     return SqlStdOperatorTable.AS.createCall(POS, asOperands);
   }
+
+  /**
+   * Override this method to handle the conversion for RelNode `f(x).y.z` where `f` is an operator, which
+   * returns a struct containing field `y`, `y` is also a struct containing field `z`.
+   *
+   * Calcite will convert this RelNode to a SqlIdentifier directly (check
+   * {@link org.apache.calcite.rel.rel2sql.SqlImplementor.Context#toSql(RexProgram, RexNode)}),
+   * which is not aligned with our expectation since we want to apply transformations on `f(x)` with
+   * {@link com.linkedin.coral.common.transformers.SqlCallTransformer}. Therefore, we override this
+   * method to convert `f(x)` to SqlCall, `.` to {@link com.linkedin.coral.common.functions.FunctionFieldReferenceOperator#DOT}
+   */
+  @Override
+  public Context aliasContext(Map<String, RelDataType> aliases, boolean qualified) {
+    return new AliasContext(INSTANCE, aliases, qualified) {
+      @Override
+      public SqlNode toSql(RexProgram program, RexNode rex) {
+        if (rex.getKind() == SqlKind.FIELD_ACCESS) {
+          final List<String> accessNames = new ArrayList<>();
+          RexNode referencedExpr = rex;
+          // Use the loop to get the top-level struct (`f(x)` in the example above),
+          // and store the accessed field names ([`z`, `y`] in the example above, needs to be reversed)
+          while (referencedExpr.getKind() == SqlKind.FIELD_ACCESS) {
+            accessNames.add(((RexFieldAccess) referencedExpr).getField().getName());
+            referencedExpr = ((RexFieldAccess) referencedExpr).getReferenceExpr();
+          }
+          if (referencedExpr.getKind() == SqlKind.OTHER_FUNCTION || referencedExpr.getKind() == SqlKind.CAST) {
```

Comment:
Also, lets look at the more recent implementation here https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java#L627 which looks different. 
If this suits out requirements, we can copy paste it and then just call out that porting back a more recent commit.

Reply:
(1) Didn't add `ROW` operator because Coral is not using this kind of operator, but I added it back to be aligned with the upstream
(2) The current upstream impl is still converting the whole RexFieldAccess to a SqlIdentifier, so backporting that doesn't meet our requirement, we still needs to override this method.